loader/src/rpc_loader.rs

```diff
@@ -699,22 +735,89 @@ fn process_load_states(
     }
 }
 
+fn process_asset_changes(
+    data: &mut LoaderData,
+    rpc: &mut RpcState,
+    asset_storage: &dyn AssetStorage,
+) -> Result<(), Box<dyn Error>> {
+    if data.pending_hot_reloads.is_empty() {
+        let changes = rpc.check_asset_changes();
+        if let Some(changes) = changes {
+            // TODO handle deleted assets
+            for asset_id in changes.changed.iter() {
+                let current_version = data
+                    .uuid_to_load
+                    .get(asset_id)
+                    .map(|l| *l)
+                    .and_then(|load_handle| {
+                        data.load_states
+                            .get(&load_handle)
+                            .map(|load| (load_handle, load))
+                    })
+                    .filter(|(_, load)| load.requested_version.is_some())
+                    .map(|(load_handle, load)| (load_handle, load.requested_version.unwrap()));
+                if let Some((handle, current_version)) = current_version {
+                    let mut load = data.load_states.get_mut(&handle).unwrap();
+                    load.pending_reload = true;
+                    data.pending_hot_reloads.push((*asset_id, current_version));
+                }
+            }
+        }
+    } else {
+        let is_finished = data.pending_hot_reloads.iter().all(|(asset_id, version)| {
+            data.uuid_to_load
+                .get(asset_id)
+                .as_ref()
+                .and_then(|load_handle| data.load_states.get(load_handle))
+                .map(|load| match load.state {
+                    LoadState::Loaded(asset_state) | LoadState::LoadingAsset(asset_state) => {
+                        match asset_state {
+                            AssetLoadState::Loaded | AssetLoadState::LoadedUncommitted => {
+                                load.requested_version.unwrap() > *version
+                            }
+                            _ => false,
+                        }
+                    }
+                    LoadState::None => true,
+                    _ => false,
+                })
+                .unwrap_or(true)
```

Comment:
hm, this would happen if:

* `pending_hot_reloads` has an asset id and `uuid_to_load` does not have that asset id, or
* `load_states` doesn't have the `load_state` for the `load_handle`.

When are these possible (if at all)?

First case, `pending_hot_reloads` is empty and an asset changed, and `uuid_to_load` does not have that asset id. Asset IDs are never removed from `uuid_to_load` (can't find any `remove` calls), and `pending_hot_reloads` can only be pushed with for an asset id from `uuid_to_load`, so this case should be impossible.

Second case, `load_handle` is removed from `load_states` when there are no more references to it.

So, if we have a pending hot reload for a non-referenced asset, we can still be considered finished. ah I think this is the thing you posted in discord.

Reply:
Yes, as you stated at the bottom of this comment, if we have a pending reload for something that is not supposed to be loaded anyway (like it does not have a LoadState), then we consider the reload to be finished/irrelevant.